### PR TITLE
Try out submitting multiple plans at once

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR opens the possibility to submit multiple plans at once.

### CODE REVIEW

**Review by jadalilleboe — CHANGES_REQUESTED** [review]

File: skyportal/handlers/api/instrument.py [hunk 1]
```diff
@@ -613,8 +613,10 @@ def put(self, instrument_id):
         self.verify_and_commit()
 
         if field_data is not None:
-            if field_region is None:
-                return self.error('field_region is required with field_data')
+            if (field_region is None) and (field_fov_type is None):
+                return self.error(
+                    'field_region or field_fov_type is required with field_data'
+                )
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/instrument.py:619]: It should be one of thus the and on the None.

File: static/js/components/ObservationPlanRequestForm.jsx [hunk 2]
```diff
@@ -235,22 +244,55 @@ const ObservationPlanRequestForm = ({ gcnevent }) => {
...
+      </div>
+      <div>
+        <Button
+          size="small"
+          color="primary"
+          type="submit"
+          variant="outlined"
+          onClick={handleSubmit}
+        >
+          Generate Observation Plans
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply static/js/components/ObservationPlanRequestForm.jsx:298]: Good point, added something better here.

File: skyportal/handlers/api/observation_plan.py [hunk 3]
```diff
@@ -870,21 +875,24 @@ def get(self, observation_plan_request_id):
...
-                continue
+        if len(observation_plan_request.observation_plans) > 0:
+            observation_plan = observation_plan_request.observation_plans[0]
+            # features are JSON representations that the d3 stuff understands.
+            # We use these to render the contours of the sky localization and
+            # locations of the transients.
+
+            geojson = []
+            fields_in = []
+            for ii, observation in enumerate(observation_plan.planned_observations):
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py]: good point.

File: static/js/components/ObservationPlanRequestForm.jsx [hunk 4]
```diff
@@ -235,22 +244,55 @@ const ObservationPlanRequestForm = ({ gcnevent }) => {
...
+        </div>
+        {isSubmitting && (
+          <div className={classes.marginTop}>
+            <CircularProgress />
+          </div>
+        )}
+      </div>
+      <div>
+        {planQueues?.map((plan) => (
+          <Button
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply static/js/components/ObservationPlanRequestForm.jsx]: Yeah you are right.

File: skyportal/handlers/api/observation_plan.py [hunk 5]
```diff
@@ -870,21 +875,24 @@ def get(self, observation_plan_request_id):
...
+                if observation.field_id not in fields_in:
+                    fields_in.append(observation.field_id)
+                    geojson.append(observation.field.contour_summary)
+                else:
+                    continue
 
-        return self.success(data={'geojson': geojson})
+            return self.success(data={'geojson': geojson})
+        else:
+            return self.error('Observation plan not yet available.')
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py:895]: What do you think would be better here?

[the remark this answers is not in the document]
Comment (jadalilleboe) [reply skyportal/handlers/api/observation_plan.py:895]: The error message is fine, just wanted to point out the buggy behavior i.e
Screen Shot 2022-04-04 at 9 50 59 PM

[the remark this answers is not in the document]
Comment (jadalilleboe) [reply skyportal/handlers/api/observation_plan.py:895]: and requests are being made over and over again forever after that.

[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py:895]: Maybe we need some kind of timeout on the LocalizationPlot looking for the geojson fields?

[the remark this answers is not in the document]
Comment (jadalilleboe) [reply skyportal/handlers/api/observation_plan.py:895]: maybe, and also maybe prevent users from adding a bunch of duplicate observation requests to the queue. I think that is also a source of errors.

[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py:895]: If you know how to do it off hand, feel free to push to my branch :)

[the remark this answers is not in the document]
Comment (jadalilleboe) [reply skyportal/handlers/api/observation_plan.py:895]: I don't, I could take a look tomorrow though. I don't really know why the infinite repetition of requests is happening

[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py:895]: Indeed, we shall both be confused!

**Comment by jadalilleboe** [comment]
The little skymaps are tweaking after these changes too
ezgif com-gif-maker (7)

**Comment by mcoughlin** [comment]
@jadalilleboe I wonder if its the attempts to prevent crashing by checking for the arrays?

**Comment by jadalilleboe** [comment]
@mcoughlin I don't think so, that should be fine. I would guess it's a backend issue

**Comment by jadalilleboe** [comment]
Actually, the issue is probably these requests that are being made over and over again.
Screen Shot 2022-04-05 at 12 30 02 PM

**Comment by mcoughlin** [comment]
@jadalilleboe So maybe a missing dependency in the useEffect(s)?

**Comment by jadalilleboe** [comment]
uh. I am not sure.

**Comment by jadalilleboe** [comment]
[https://stackoverflow.com/questions/53949393/cant-perform-a-react-state-update-on-an-unmounted-component](https://stackoverflow.com/questions/53949393/cant-perform-a-react-state-update-on-an-unmounted-component) this looks promising!

**Comment by mcoughlin** [comment]
@jadalilleboe this isMounted bit isn't obviously working, i.e. it wont query at all! :)

File: skyportal/handlers/api/observation_plan.py [hunk 6]
```diff
@@ -71,72 +71,77 @@ def post(self):
                               description: New observation plan request ID
         """
         data = self.get_json()
```
Comment (guynir42) [inline skyportal/handlers/api/observation_plan.py:73]: maybe change the first instance of `data` to `json`.

File: skyportal/handlers/api/observation_plan.py [hunk 7]
```diff
@@ -71,72 +71,77 @@ def post(self):
...
+            data['localization_id'] = int(data['localization_id'])
 
-        if not instrument.api_class_obsplan.implements()['submit']:
-            return self.error(
-                'Cannot submit observation plan requests for this Instrument.'
+            allocation = Allocation.get_if_accessible_by(
+                data['allocation_id'],
+                self.current_user,
+                raise_if_none=True,
             )
```
Comment (guynir42) [inline skyportal/handlers/api/observation_plan.py:96]: catch this if allocation isn't found.

File: skyportal/handlers/api/observation_plan.py [hunk 8]
```diff
@@ -71,72 +71,77 @@ def post(self):
...
+                )
 
-        # validate the payload
-        jsonschema.validate(data['payload'], formSchema)
+            target_groups = []
+            for group_id in data.pop('target_group_ids', []):
+                g = Group.get_if_accessible_by(
+                    group_id, self.current_user, raise_if_none=True
+                )
+                target_groups.append(g)
```
Comment (guynir42) [inline skyportal/handlers/api/observation_plan.py:112]: catch this, too.

File: skyportal/handlers/api/observation_plan.py [hunk 9]
```diff
@@ -71,72 +71,77 @@ def post(self):
...
+                )
+            except AttributeError:
+                formSchema = instrument.api_class_obsplan.form_json_schema
 
-        self.push_all(
-            action="skyportal/REFRESH_GCNEVENT",
-            payload={"gcnEvent_dateobs": observation_plan_request.gcnevent.dateobs},
-        )
+            # validate the payload
+            jsonschema.validate(data['payload'], formSchema)
```
Comment (guynir42) [inline skyportal/handlers/api/observation_plan.py:122]: what happens if validation fails?

**Review by guynir42 — APPROVED** [review]
Looks good overall. Need to catch some exceptions in a few places. Also, make sure to figure out the bug that Jada found with the infinite loop before merging this.

File: skyportal/handlers/api/observation_plan.py (hunk 9 above)
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py:122]: catching this.

File: skyportal/handlers/api/observation_plan.py (hunk 8 above)
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py:112]: yes

File: skyportal/handlers/api/observation_plan.py (hunk 7 above)
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py:96]: yes

File: skyportal/handlers/api/observation_plan.py (hunk 6 above)
Comment (mcoughlin) [reply skyportal/handlers/api/observation_plan.py:73]: good idea.

**Comment by mcoughlin** [comment]
@jadalilleboe there was an extra dependency in the effect that shouldn't have been there, so at least the maps render now

**Review by jadalilleboe — APPROVED** [review]

**Comment by jadalilleboe** [comment]
Maybe the "submit" button should be "submit to queue". Then "generate observation plans" could be "add observation plans for all in queue" or something. That's a long button name though, probably just "generate observation plans" or "add observation plans" would be fine. @mcoughlin

**Comment by jadalilleboe** [comment]
also, I think the "generate observation plans" or "add observation plans" button shouldn't be visible if there's no plans in the queue. Then you could get rid of the error message too if the button's clicked with none in the queue. @mcoughlin

**Comment by mcoughlin** [comment]
@jadalilleboe you can see if you like the way its done more now?

**Comment by jadalilleboe** [comment]
@mcoughlin yeah looks great!

**Pull request merged**